File: Modules/Social.py
# File = robjam1990/Psychosis/Gameplay/Planets/Thear/Social.py
# Class for managing interactions between NPCs
class InteractionManager:
    def __init__(self):
        self.interactions = []

    # Add interaction between two NPCs
    def add_interaction(self, interaction):
        self.interactions.append(interaction)

    # Resolve interactions
    def resolve_interactions(self):
        for interaction in self.interactions:
            if not interaction or not interaction.npc1 or not interaction.npc2:
                print("Invalid interaction data:", interaction)
                return
            self.resolve_interaction_type(interaction)

    def resolve_interaction_type(self, interaction):
        if interaction.type == "conversation":
            # Implement conversation resolution logic
            print(f"Conducting conversation between {interaction.npc1.identification.name} and {interaction.npc2.identification.name}")
        elif interaction.type == "combat":
            # Implement combat resolution logic
            print(f"Engaging in combat between {interaction.npc1.identification.name} and {interaction.npc2.identification.name}")
        else:
            print(f"Unhandled interaction type: {interaction.type}")

# Class representing an interaction between NPCs
class Interaction:
    def __init__(self, npc1, npc2, type):
        self.npc1 = npc1
        self.npc2 = npc2
        self.type = type

# Quest class representing tasks or missions in the game
class Quest:
    def __init__(self, description, reward):
        self.description = description
        self.reward = reward
        self.completed = False

    # Mark the quest as completed
    def complete_quest(self):
        self.completed = True

# NPC dialogue manager for handling conversations
class DialogueManager:
    def __init__(self):
        self.dialogues = []

    # Add dialogue between NPCs or NPCs and players
    def add_dialogue(self, dialogue):
        self.dialogues.append(dialogue)

    # Initiate dialogue between two NPCs or an NPC and a player
    def initiate_dialogue(self, participant1, participant2):
        # Implement dialogue initiation logic here
        print(f"Initiating dialogue between {participant1.identification.name} and {participant2.identification.name}")

# Add functionality to handle reputation with factions or groups
class ReputationSystem:
    # Existing code...

    # Update the player's reputation with a faction or group
    def update_faction_reputation(self, faction, reputation_change):
        if faction not in self.reputations:
            self.reputations[faction] = 0
        self.reputations[faction] += reputation_change

    # Get the player's reputation with a faction or group
    def get_faction_reputation(self, faction):
        return self.reputations.get(faction, 0)

# Main game class
class Game:
    def __init__(self):
        self.organisms = []
        self.bounties = []
        self.hierarchies = []
        # self.justice_system = JusticeSystem()  # You need to define JusticeSystem class
        self.locations = []
        self.interaction_manager = InteractionManager()
        self.dialogue_manager = DialogueManager()
        self.reputation_system = ReputationSystem()

    # Add organism to the game's ecosystem
    def add_organism(self, organism):
        self.organisms.append(organism)

    # Add bounty to the game
    def add_bounty(self, bounty):
        self.bounties.append(bounty)

    # Add hierarchy to the game
    def add_hierarchy(self, hierarchy):
        self.hierarchies.append(hierarchy)

    # Display all organisms in the ecosystem
    def display_organisms(self):
        for organism in self.organisms:
            print(f"Species: {organism.species}, Age: {organism.age}")

    # Example communication round
    def conduct_communication_round(self):
        for organism in self.organisms:
            organism.communicate("Hello from the game.")

    # Resolve interactions between NPCs
    def resolve_interactions(self):
        self.interaction_manager.resolve_interactions()

    # Add dialogue between NPCs or NPCs and players
    def add_dialogue(self, dialogue):
        self.dialogue_manager.add_dialogue(dialogue)

    # Initiate dialogue between two NPCs or an NPC and a player
    def initiate_dialogue(self, participant1, participant2):
        self.dialogue_manager.initiate_dialogue(participant1, participant2)

    # Update player's reputation with an NPC
    def update_reputation(self, npc, reputation_change):
        self.reputation_system.update_faction_reputation(npc, reputation_change)

    # Get player's reputation with an NPC
    def get_reputation(self, npc):
        return self.reputation_system.get_faction_reputation(npc)

# Main program code
game = Game()

# Adding organisms to the game's ecosystem
# game.add_organism(Animal("Wolf", 4))  # You need to define Animal class
# game.add_organism(Animal("Deer", 3))  # You need to define Animal class

# Adding NPCs to the game's ecosystem
# john = NPC("John", 75)  # You need to define NPC class
# jane = NPC("Jane", 60)  # You need to define NPC class
# game.add_organism(john)
# game.add_organism(jane)

# Add an interaction between two NPCs
# interaction = Interaction(john, jane, "conversation")
# game.interaction_manager.add_interaction(interaction)

# Initiate dialogue between two NPCs
# game.initiate_dialogue(john, jane)

# Resolve interactions between NPCs
# game.resolve_interactions()

# Update player's reputation with an NPC
# game.update_reputation(john, 10)

# Get player's reputation with an NPC
# john_reputation = game.get_reputation(john)
# print(f"Player's reputation with John: {john_reputation}")

File: Modules/Social.js
// File = robjam1990/Psychosis/Gameplay/Planets/Thear/Social.js
// Class for managing interactions between NPCs
class InteractionManager {
    constructor() {
        this.interactions = [];
    }

    // Add interaction between two NPCs
    addInteraction(interaction) {
        this.interactions.push(interaction);
    }

    // Resolve interactions
    resolveInteractions() {
        this.interactions.forEach(interaction => {
            if (!interaction || !interaction.npc1 || !interaction.npc2) {
                console.error("Invalid interaction data:", interaction);
                return;
            }
            this.resolveInteractionType(interaction);
        });
    }

    resolveInteractionType(interaction) {
        switch (interaction.type) {
            case "conversation":
                // Implement conversation resolution logic
                console.log(`Conducting conversation between ${interaction.npc1.identification.name} and ${interaction.npc2.identification.name}`);
                break;
            case "combat":
                // Implement combat resolution logic
                console.log(`Engaging in combat between ${interaction.npc1.identification.name} and ${interaction.npc2.identification.name}`);
                break;
            // Add more cases for other interaction types as needed
            default:
                console.log(`Unhandled interaction type: ${interaction.type}`);
        }
    }
}

// Class representing an interaction between NPCs
class Interaction {
    constructor(npc1, npc2, type) {
        this.npc1 = npc1;
        this.npc2 = npc2;
        this.type = type;
    }
}

// Quest class representing tasks or missions in the game
class Quest {
    constructor(description, reward) {
        this.description = description;
        this.reward = reward;
        this.completed = false;
    }

    // Mark the quest as completed
    completeQuest() {
        this.completed = true;
    }
}

// NPC dialogue manager for handling conversations
class DialogueManager {
    constructor() {
        this.dialogues = [];
    }

    // Add dialogue between NPCs or NPCs and players
    addDialogue(dialogue) {
        this.dialogues.push(dialogue);
    }

    // Initiate dialogue between two NPCs or an NPC and a player
    initiateDialogue(participant1, participant2) {
        // Implement dialogue initiation logic here
        console.log(`Initiating dialogue between ${participant1.identification.name} and ${participant2.identification.name}`);
    }
}

// Add functionality to handle reputation with factions or groups
class ReputationSystem {
    // Existing code...

    // Update the player's reputation with a faction or group
    updateFactionReputation(faction, reputationChange) {
        if (!(faction in this.reputations)) {
            this.reputations[faction] = 0;
        }
        this.reputations[faction] += reputationChange;
    }

    // Get the player's reputation with a faction or group
    getFactionReputation(faction) {
        return this.reputations[faction] || 0;
    }
}

// Main game class
class Game {
    constructor() {
        this.organisms = [];
        this.bounties = [];
        this.hierarchies = [];
        this.justiceSystem = new JusticeSystem();
        this.locations = [];
        this.interactionManager = new InteractionManager();
        this.dialogueManager = new DialogueManager();
        this.reputationSystem = new ReputationSystem();
    }

    // Add organism to the game's ecosystem
    addOrganism(organism) {
        this.organisms.push(organism);
    }

    // Add bounty to the game
    addBounty(bounty) {
        this.bounties.push(bounty);
    }

    // Add hierarchy to the game
    addHierarchy(hierarchy) {
        this.hierarchies.push(hierarchy);
    }

    // Display all organisms in the ecosystem
    displayOrganisms() {
        this.organisms.forEach(organism => {
            console.log(`Species: ${organism.species}, Age: ${organism.age}`);
        });
    }

    // Example communication round
    conductCommunicationRound() {
        this.organisms.forEach(organism => {
            organism.communicate("Hello from the game.");
        });
    }

    // Resolve interactions between NPCs
    resolveInteractions() {
        this.interactionManager.resolveInteractions();
    }

    // Add dialogue between NPCs or NPCs and players
    addDialogue(dialogue) {
        this.dialogueManager.addDialogue(dialogue);
    }

    // Initiate dialogue between two NPCs or an NPC and a player
    initiateDialogue(participant1, participant2) {
        this.dialogueManager.initiateDialogue(participant1, participant2);
    }

    // Update player's reputation with an NPC
    updateReputation(npc, reputationChange) {
        this.reputationSystem.updateReputation(npc, reputationChange);
    }

    // Get player's reputation with an NPC
    getReputation(npc) {
        return this.reputationSystem.getReputation(npc);
    }
}

// Main program code
const game = new Game();

// Adding organisms to the game's ecosystem
game.addOrganism(new Animal("Wolf", 4));
game.addOrganism(new Animal("Deer", 3));

// Adding NPCs to the game's ecosystem
const john = new NPC("John", 75);
const jane = new NPC("Jane", 60);
game.addOrganism(john);
game.addOrganism(jane);

// Add an interaction between two NPCs
const interaction = new Interaction(john, jane, "conversation");
game.interactionManager.addInteraction(interaction);

// Initiate dialogue between two NPCs
game.initiateDialogue(john, jane);

// Resolve interactions between NPCs
game.resolveInteractions();

// Update player's reputation with an NPC
game.updateReputation(john, 10);

// Get player's reputation with an NPC
const johnReputation = game.getReputation(john);
console.log(`Player's reputation with John: ${johnReputation}`);
